spec/support/delayed_job.rb

```diff
@@ -15,3 +15,28 @@ class TestCronJob < CronJob
 
   def perform; end
 end
+
+module DelayedJobMatchers
+  extend RSpec::Matchers::DSL
+
+  define :delay_execution_of do |method_name|
+    match do |actual|
+      handler = Delayed::Job.new(payload_object: Delayed::PerformableMethod.new(actual, method_name, [])).handler
+      jobs = Delayed::Job.where("handler LIKE ?", handler.lines.first(3).join + "%")
+
+      return jobs.any? unless @arguments
+
+      jobs.any? do |job|
+        @arguments.args_match?(*job.payload_object.args)
+      end
+    end
+
+    chain :with do |*args|
+      @arguments = RSpec::Mocks::ArgumentListMatcher.new(*args)
+    end
+  end
```

Comment:
This is all pretty hard to understand to me - and I imagine to anyone who doesn't go into the details of classes we extend. Obvious questions being 'why do we ned the first three lines of handler in the query?', 'where do @arguments come from', 'what the hell is chain for', 'is handler the best place to check it', and I'm sure there are others.

Why do we need this level of power and precision in our tests? Are there any tests where we need to differentiate between the names of the jobs? Do we ever need to check the arguments of scheduled jobs? Is there no way to do this level of precision with specific stubs, rather than generic matchers?

Reply:
@kicferk1 Absolutely agree, this is not the easiest code. This is to deal with a limitation of ActiveJob abstraction. What we need to test is that given job/process has been enqueued, and there is no way of testing this using neither ActiveJob test helpers nor stubs as they are both unaware of DelayedJob transactional behaviour. 

On the upside - this is strictly technical domain. There is absolutely no business logic in this code and as such it is the best place for complexity to live in, while at the same time it makes the test code much simpler and easy to follow.

To answer other question: `chain` is part of RSpec::Matchers::DSL and is pretty common tool for defining new matchers. It defines a new method on the matcher which executes the code in the block and then returns itself. It is an equivalent of:

```
def with(*args)
  @arguments = ...
  self
end
```

and is used so the matcher can be used like this:

```
expect(object).to my_matcher_name.with(....)
```

`@arguments` are defined within optional `with` call.

Reply:
**Are there any tests where we need to differentiate between the names of the jobs?**
We need to be confident we are en-queuing the correct job. ActiveJob test helpers work in a very similar manner - `enqueued_jobs` is just an array of hashes containg the name ofthe job and its arguments.

Reply:
Ech, this looks like classic `YAGNI` to me (You Aint Gonna Need It). Particular the part for defining `.with` method. Then again, the hard work for it has been done, so we might as well use it...

But if we do, I want it to be more readable.

I would put `chain` definition above `match` definition, to signify which one is happening first (cause I think `chain` precedes `match`, right?). 

Then inside `match`, let's make the line 
```
jobs = Delayed::Job.where("handler LIKE ?", handler.lines.first(3).join + "%")
```
a bit less magic - let's define the query more explicitly, and name the first three lines of our handler. 

I wonder if we need defining an in-memory handler too, maybe we could fetch the handlers from database using the method names we already know. But that should become more apparent when we have a more explicit query in place.

Reply:
I'd argue `YAGNI` does not apply here (and in does not apply in 95% of places when it is being used). There is an actual test which could not be written without this matcher - namely that no job is scheduled if a single partnership object fails to save. Also `with` is used to ensure that the partnership mailer has been enqueued with the correct arguments.

I am not sure how changing the order of match and chain would help, both those calls are DSL methods to define methods on a matcher and their execution does not matter. Morover, `with` call is optional. This is a standard matcher definition as per Rspec documentaiton: https://relishapp.com/rspec/rspec-expectations/v/3-10/docs/custom-matchers/define-matcher-with-fluent-interface.

As per the line suggested - it is not possible to write it with less magic. DelayedJob is serializing objects into yaml, of which the first 3 lines define the class of the object receiving the call. The goal of this query is to limit a number of jobs received to a reasonable amount - alternatively we could just loop over all the jobs, serialize all the payload_objects in search for an exact match. Another option is to configure DelayedJobs to use json columns instead and use json operators in the query.

Reply:
> As per the line suggested - it is not possible to write it with less magic. DelayedJob is serializing objects into yaml, of which the first 3 lines define the class of the object receiving the call.

Fine, let's literally have three variables, one for each of those lines. So someone reading the code knows _why_ we care about first three lines, and what they represent. So they don't think 'hmm, I wonder if I can reduce it to first two lines, or just one line out of those three'.

Reply:
@kicferk1 There's really no way to name those variable which would make much sense. This is an example of a handler as DJ stores it in the database:

```
--- !ruby/object:Delayed::PerformableMethod
object: !ruby/object:SchoolDataImporter
  logger: !ruby/object:ActiveSupport::Logger
    level: 0
    progname: 
    default_formatter: !ruby/object:Logger::Formatter
      datetime_format: 
    formatter: !ruby/object:ActiveSupport::Logger::SimpleFormatter
      datetime_format: 
      thread_key: activesupport_tagged_logging_tags:17500
    logdev: !ruby/object:Logger::LogDevice
      shift_period_suffix: "%Y%m%d"
      shift_size: 1048576
      shift_age: 0
      filename: "/home/broisatse/work/dfe/early-careers-framework/log/development.log"
      dev: !ruby/object:File {}
      binmode: false
      mon_data: !ruby/object:Monitor {}
      mon_data_owner_object_id: 6640
  start_year: 2021
method_name: :run
args: []
```

PerformableObject is a object defined by DelayedJob, which represents `.delay` call. It encapsulates 3 piece of information, receiving object, method to delay and arguments. In ideal world we would grab yaml representation of the entire `object` attribute, however it might be time dependent which is why I have only grabbed the first 3 lines.

I really see no value in assigning it into 3 separate variables - the whole goal of this query is to get all the jobs that has been enqueued for objects of the same class as object given in `expect`. This code cannot be understood without knowing the internals of DelayedJob, even with additional variables.

I'll defenitely add some comments explaining how this behaves, however if anyone was to apporach this in the future, short `byebug` session would make everything clear.

Reply:
Right, so the `3` is arbitrary. I got the impression that the number was specifically important.

I wonder if the real thing we are looking for is the line `object: !ruby/object:XXX`? Maybe we could call it something like `handler_object_class_line` and get objects based off that? Is that line 2?